File: desktop/src/websocket/ApiClient.js
'use strict';

/**
 * Thin REST client for the Clear backend.
 * Holds the access token in memory, keeps the refresh token in SettingsStore
 * (encrypted at rest), and transparently refreshes on 401.
 */

class ApiError extends Error {
  constructor(message, { status, code, body } = {}) {
    super(message);
    this.name = 'ApiError';
    this.status = status || 0;
    this.code = code || 'request_failed';
    this.body = body;
  }
}

class ApiClient {
  constructor({ settings, logger }) {
    this.settings = settings;
    this.log = logger;
    this.accessToken = null;
    this.user = null;
    this.refreshing = null;
  }

  get baseUrl() {
    return String(this.settings.get('backendUrl') || '').replace(/\/+$/, '');
  }

  get refreshToken() {
    return this.settings.getSecret('refreshToken');
  }

  get signedIn() {
    return Boolean(this.refreshToken || this.accessToken);
  }

  async request(path, { method = 'GET', body, auth = true, retryOn401 = true, timeoutMs = 15000 } = {}) {
    if (!this.baseUrl) throw new ApiError('No backend URL configured', { code: 'no_backend' });

    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), timeoutMs);

    try {
      const response = await fetch(`${this.baseUrl}${path}`, {
        method,
        headers: {
          'content-type': 'application/json',
          ...(auth && this.accessToken ? { authorization: `Bearer ${this.accessToken}` } : {}),
          ...(this.settings.get('device.id') ? { 'x-device-id': this.settings.get('device.id') } : {}),
        },
        body: body ? JSON.stringify(body) : undefined,
        signal: controller.signal,
      });

      const text = await response.text();
      const json = text ? safeParse(text) : null;

      if (response.status === 401 && auth && retryOn401 && this.refreshToken) {
        await this.refresh();
        return this.request(path, { method, body, auth, retryOn401: false, timeoutMs });
      }

      if (!response.ok) {
        throw new ApiError(json?.message || `Request failed (${response.status})`, {
          status: response.status,
          code: json?.error,
          body: json,
        });
      }

      return json;
    } catch (error) {
      if (error.name === 'AbortError') throw new ApiError('The backend did not respond in time', { code: 'timeout' });
      if (error instanceof ApiError) throw error;
      throw new ApiError(error.message || 'Network error', { code: 'network' });
    } finally {
      clearTimeout(timer);
    }
  }

  async login({ email, password }) {
    const result = await this.request('/login', {
      method: 'POST',
      auth: false,
      body: {
        email,
        password,
        platform: 'desktop',
        deviceId: this.settings.get('device.id'),
        deviceName: this.settings.get('device.name'),
      },
    });

    this.accessToken = result.accessToken;
    this.user = result.user;
    this.settings.patch({
      refreshToken: result.refreshToken,
      auth: { email: result.user.email, userId: result.user.id },
    });

    this.log?.info('Signed in', { userId: result.user.id });
    return result;
  }

  /** Rotates the refresh token; concurrent callers share one in-flight request. */
  async refresh() {
    if (this.refreshing) return this.refreshing;

    const token = this.refreshToken;
    if (!token) throw new ApiError('Not signed in', { status: 401, code: 'no_refresh_token' });

    this.refreshing = (async () => {
      try {
        const result = await this.request('/auth/refresh', {
          method: 'POST',
          auth: false,
          body: { refreshToken: token },
        });
        this.accessToken = result.accessToken;
        this.user = result.user;
        this.settings.patch({
          refreshToken: result.refreshToken,
          auth: { email: result.user.email, userId: result.user.id },
        });
        this.log?.debug('Access token refreshed');
        return result;
      } catch (error) {
        if (error.status === 401) {
          this.log?.warn('Refresh token rejected - signing out');
          this.signOutLocal();
        }
        throw error;
      } finally {
        this.refreshing = null;
      }
    })();

    return this.refreshing;
  }

  /** Ensures a usable access token, refreshing if we only have the long-lived one. */
  async ensureToken() {
    if (this.accessToken) return this.accessToken;
    if (!this.refreshToken) throw new ApiError('Not signed in', { status: 401, code: 'signed_out' });
    const result = await this.refresh();
    return result.accessToken;
  }

  async logout() {
    try {
      if (this.accessToken) {
        await this.request('/auth/logout', { method: 'POST', body: { refreshToken: this.refreshToken } });
      }
    } catch (error) {
      this.log?.debug('Logout call failed (ignoring)', { error: error.message });
    }
    this.signOutLocal();
  }

  signOutLocal() {
    this.accessToken = null;
    this.user = null;
    this.settings.patch({ refreshToken: null, auth: { email: null, userId: null } });
  }

  createPairingCode() {
    return this.request('/pair/code', {
      method: 'POST',
      body: {
        deviceId: this.settings.get('device.id'),
        deviceName: this.settings.get('device.name'),
      },
    });
  }

  listDevices() {
    return this.request('/pair/devices');
  }

  history({ limit = 30, search } = {}) {
    const params = new URLSearchParams({ limit: String(limit) });
    if (search) params.set('search', search);
    return this.request(`/history?${params.toString()}`);
  }

  postAnswer(payload) {
    return this.request('/answer', { method: 'POST', body: payload });
  }

  health() {
    return this.request('/health', { auth: false, timeoutMs: 6000 });
  }
}

const safeParse = (text) => {
  try {
    return JSON.parse(text);
  } catch {
    return { message: text.slice(0, 200) };
  }
};

module.exports = { ApiClient, ApiError };

File: mobile/lib/data/api_client.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';

import 'package:http/http.dart' as http;

import '../core/config.dart';
import '../core/storage.dart';
import 'models.dart';

/// REST layer. Keeps the access token in memory, the refresh token in the
/// keystore, and refreshes transparently on a 401.
class ApiClient {
  ApiClient(this._storage);

  final Storage _storage;
  final http.Client _http = http.Client();

  String? _accessToken;
  Future<void>? _refreshing;

  String get baseUrl => _storage.backendUrl;

  Uri _uri(String path, [Map<String, String>? query]) =>
      Uri.parse('$baseUrl$path').replace(queryParameters: query?.isEmpty ?? true ? null : query);

  Map<String, String> _headers({bool auth = true}) => {
        'content-type': 'application/json',
        'x-device-id': _storage.deviceId,
        if (auth && _accessToken != null) 'authorization': 'Bearer $_accessToken',
      };

  Future<Map<String, dynamic>> _send(
    String method,
    String path, {
    Map<String, dynamic>? body,
    Map<String, String>? query,
    bool auth = true,
    bool retryOn401 = true,
  }) async {
    try {
      final request = http.Request(method, _uri(path, query))
        ..headers.addAll(_headers(auth: auth));
      if (body != null) request.body = jsonEncode(body);

      final streamed = await _http.send(request).timeout(AppConfig.requestTimeout);
      final response = await http.Response.fromStream(streamed);

      if (response.statusCode == 401 && auth && retryOn401) {
        await refresh();
        return _send(method, path, body: body, query: query, auth: auth, retryOn401: false);
      }

      final decoded = response.body.isEmpty ? <String, dynamic>{} : jsonDecode(response.body);
      final map = decoded is Map<String, dynamic> ? decoded : <String, dynamic>{'data': decoded};

      if (response.statusCode >= 400) {
        throw ApiException(
          (map['message'] ?? 'Request failed (${response.statusCode})').toString(),
          statusCode: response.statusCode,
          code: map['error']?.toString(),
        );
      }
      return map;
    } on SocketException {
      throw ApiException('Cannot reach the server. Check the backend URL and your connection.');
    } on TimeoutException {
      throw ApiException('The server took too long to respond.');
    } on FormatException {
      throw ApiException('The server returned an unexpected response.');
    }
  }

  // ---- auth ---------------------------------------------------------------

  Future<UserSession> login({required String email, required String password}) async {
    final result = await _send('POST', '/login', auth: false, body: {
      'email': email,
      'password': password,
      'platform': 'mobile',
      'deviceId': _storage.deviceId,
      'deviceName': 'Android phone',
    });

    final session = UserSession.fromJson(result);
    _accessToken = session.accessToken;
    await _storage.writeRefreshToken(session.refreshToken);
    await _storage.setEmail(session.email);
    await _storage.setUserId(session.userId);
    return session;
  }

  /// Rotates the refresh token. Concurrent callers await the same future.
  Future<String> refresh() async {
    if (_refreshing != null) {
      await _refreshing;
      return _accessToken ?? (throw ApiException('Session expired', statusCode: 401));
    }

    final completer = Completer<void>();
    _refreshing = completer.future;

    try {
      final token = await _storage.readRefreshToken();
      if (token == null) throw ApiException('Signed out', statusCode: 401);

      final result = await _send('POST', '/auth/refresh', auth: false, retryOn401: false, body: {
        'refreshToken': token,
      });

      final session = UserSession.fromJson(result);
      _accessToken = session.accessToken;
      await _storage.writeRefreshToken(session.refreshToken);
      await _storage.setEmail(session.email);
      await _storage.setUserId(session.userId);
      completer.complete();
      return session.accessToken;
    } on ApiException catch (error) {
      completer.complete();
      if (error.isAuthError) await _storage.clearSession();
      rethrow;
    } finally {
      _refreshing = null;
    }
  }

  /// Returns a usable access token, refreshing if needed.
  Future<String> ensureToken() async => _accessToken ?? await refresh();

  Future<void> logout() async {
    try {
      final token = await _storage.readRefreshToken();
      await _send('POST', '/auth/logout', body: {'refreshToken': token});
    } catch (_) {
      // Signing out locally matters more than telling the server.
    }
    _accessToken = null;
    await _storage.clearSession();
  }

  // ---- pairing ------------------------------------------------------------

  Future<Map<String, dynamic>> pair(String code) => _send('POST', '/pair', body: {
        'code': code.trim().toUpperCase(),
        'deviceId': _storage.deviceId,
        'deviceName': 'Android phone',
      });

  Future<List<Map<String, dynamic>>> devices() async {
    final result = await _send('GET', '/pair/devices');
    return (result['devices'] as List? ?? []).cast<Map<String, dynamic>>();
  }

  // ---- history ------------------------------------------------------------

  Future<List<Answer>> history({int limit = AppConfig.historyPageSize, int? before, String? search}) async {
    final result = await _send('GET', '/history', query: {
      'limit': '$limit',
      if (before != null) 'before': '$before',
      if (search != null && search.isNotEmpty) 'search': search,
    });
    return (result['answers'] as List? ?? [])
        .map((json) => Answer.fromJson(json as Map<String, dynamic>))
        .toList();
  }

  Future<Map<String, dynamic>> health() => _send('GET', '/health', auth: false);

  void dispose() => _http.close();
}
